### ISSUE TITLE
Optional/Default arguments for a tcl proc in Rust code?

### ISSUE THREAD

**Issue by benburkhart1** [oooutlk/tcltk#7]
How can you have a default or optional argument for a Tcl proc within Rust code?

Trying for instance using tclosure!

```
          let cmd = tclosure!( &interp, cmd: "testproc",
              move |message: String, options: String| -> TclResult<i32> {
                  Ok(1)
              });

           // I'm aware I should be using cmd here
           let c = interp.eval("testproc \"foo\"");
```

Results in:
```
Err(wrong # args: should be "testproc"
    invoked from within
```

I've tried using a tcl way of providing a default in the "args", but the parameter is not really documented at all, and you don't give a non empty example in the documentation.
```
            let cmd = tclosure!( &interp, cmd: "testproc", args: "message {options \"bar\"}",
              move |message: String, options: String| -> TclResult<i32> {
                  Ok(1)
              });
```

Which also does not work. How should I be doing this using your library?

**Comment by benburkhart1** [comment]
Here's the TCL version of what I'm trying to accomplish:
```

proc testproc { message { options "bar" } } {
    puts $message
    puts $options
}

testproc "foo"
```


Results in:
```
foo
bar
```

**Comment by oooutlk** [comment]
Default arguments are supported in Tcl but not in Rust nor in tcl_derive crate. If it were supported, the syntax may be like the following:

```
let cmd = tclosure!( interp,
    cmd: "testproc",
    move |message: String, options: String = String::from("bar") | -> TclResult<i32> {
        Ok(1)
    }
);
```

I hesitated to support this because it introduces a default argument syntax which does not exist in Rust and may cause misleading. But it's possible.

By the way, the "args" argument in `tclosure!()` has nothing to do with the command argument **definitions**. It's all about providing argouments for command **invocation**. So your code `args: "message {options \"bar\"}"` just makes `cmd` ( in `let cmd =` ) a command invocation `testproc message {options "bar"}`.

An [example](https://oooutlk.github.io/tk/event_handling.html#binding-to-events)
to help understanding `args` in `tclosure!()`:

```
    l.bind( event::button_3().motion(), tclosure!( tk, args: "%x %y hello",
        move |x: i32, y: i32, _z: Obj| -> TkResult<()> {
            Ok( l.configure( -text( format!( "right button drag to {} {}", x, y )))? )
        }
    ))?;
```

The `tclosure!()` macro defines a Tcl command with three arguments, and returns an Tcl Obj to invoke this very command, with the arguments %x, %y and "hello".

**Comment by oooutlk** [comment]
Another syntax: we can extend the `cmd` parameter in `tclosure!()` to encode the Tcl command argument lists. It’s much easier to implement. For example, a `cmd: “testproc { message { options \"bar\" }”` will do what you want.

**Comment by oooutlk** [comment]
Yet another syntax:

```
tclosure!( interp,
    cmd: "testproc",
    move |message: String, #[default("bar")] options: String| -> TclResult<i32> {
        Ok(1)
    }
);
```

**Comment by oooutlk** [comment]
This feature has been implemented in [tcl_derive 0.1.2](https://crates.io/crates/tcl_derive/0.1.2).

See the [test case](https://github.com/oooutlk/tcltk/blob/49fc21844acbb07738d6fa12931cf0468500936c/tcl/src/interp.rs#L723) for demonstration.
